src/websocket/server.cc

```diff
@@ -166,13 +166,97 @@ future<> connection::read_http_upgrade_request() {
+                    _cstate = connection_state::error;
+                    return websocket_parser::stop(std::move(data));
+                }
+            }
+            _state = parsing_state::payload_length_and_mask;
+        } else {
+            _buffer.append(data.get(), data.size());
+            return websocket_parser::dont_stop();
+        }
+    }
+    if (_state == parsing_state::payload_length_and_mask) {
+        size_t const required_bytes = _header->get_rest_of_header_length();
+        if (_buffer.length() + data.size() >= required_bytes) {
+            if (_buffer.length() < required_bytes) {
+                size_t hlen = _buffer.length();
+                _buffer.append(data.get(), required_bytes - hlen);
+                data.trim_front(required_bytes - hlen);
+
+                _payload_length = _header->length;
+                size_t offset = 0;
+                char const *input = _buffer.data();
+                if (_header->length == 126) {
+                    _payload_length = be16toh(*(uint16_t const *)(input + offset));
+                    offset += sizeof(uint16_t);
+                } else if (_header->length == 127) {
+                    _payload_length = be64toh(*(uint64_t const *)(input + offset));
+                    offset += sizeof(uint64_t);
+                }
+                _masking_key = be32toh(*(uint32_t const *)(input + offset));
+                _buffer = {};
+            }                
+            _state = parsing_state::payload;
+        } else {
+            _buffer.append(data.get(), data.size());
+            return websocket_parser::dont_stop();
+        }
+    }
+    if (_state == parsing_state::payload) {
+        if (_payload_length > data.size()) {
+            _payload_length -= data.size();
+            remove_mask(data, data.size());
+            _result = std::move(data);
+            return websocket_parser::stop(buff_t(0));
+        } else { 
+            _result = data.clone();
+            remove_mask(_result, _payload_length);
+            data.trim_front(_payload_length);
+            _payload_length = 0;
+            _state = parsing_state::flags_and_payload_data;
+            return websocket_parser::stop(std::move(data));
+        }
+    }
+    _cstate = connection_state::error;
+    return websocket_parser::stop(std::move(data));
+}
+
 future<> connection::read_one() {
-    return _read_buf.read().then([this] (temporary_buffer<char> buf) {
-        if (buf.empty()) {
+    return _read_buf.consume(_websocket_parser).then([this] () mutable {
+        if (_websocket_parser.is_valid()) {
+            return write_to_pipe(_websocket_parser.result());
+        } else if (_websocket_parser.eof()) {
             _done = true;
+            return make_ready_future<>();    
         }
-        //FIXME: implement
-        wlogger.info("Received: {}", buf.get());
+        // ERROR
+        wlogger.error("Reading from socket has failed.");
```

Comment:
I don't think a library should print something when read failed - it's probably a fairly standard occurrence when the client had a problem, isn't it?
Maybe some sort of debugging message, but probably not an "error"?
I have a similar question on other wlogger.error() calls below.